pkg/ebpf/c/tracer-ebpf.c

```diff
@@ -845,6 +906,176 @@ int kprobe__tcp_v4_destroy_sock(struct pt_regs* ctx) {
+    __u64 tid = bpf_get_current_pid_tgid();
+
+    // determine if the fd for this process is an unbound UDP socket
+    __u64 fd_and_tid = (tid << 32) | fd;
+    __u64* u = bpf_map_lookup_elem(&unbound_sockets, &fd_and_tid);
+
+    if (u == NULL) {
+        log_debug("kprobe/sys_bind: bind happened, but not on a UDP socket\n");
+        return 0;
+    }
+
+    struct sockaddr* addr = (struct sockaddr*)PT_REGS_PARM2(ctx);
+    if (addr == NULL) {
+        return 0;
+    }
+
+    // sockaddr is part of the syscall ABI, so we can hardcode the offset of 2 to find the port.
+    u16 sin_port = 0;
+    bpf_probe_read(&sin_port, sizeof(u16), (char*)addr + 2);
+    sin_port = ntohs(sin_port);
+
+    // write to pending_binds to the retprobe knows we can mark this as binding.
+    bind_syscall_args_t args = {};
+    args.fd = fd;
+    args.port = sin_port;
+
+    bpf_map_update_elem(&pending_bind, &tid, &args, BPF_ANY);
+    log_debug("kprobe/sys_bind: started a bind on UDP port=%d fd=%d tid=%d\n", sin_port, fd, tid);
+
+
+
+    return 0;
+}
+
+SEC("kretprobe/sys_bind")
+int kretprobe__sys_bind(struct pt_regs* ctx) {
+  __u64 tid = bpf_get_current_pid_tgid();
+  int ret = PT_REGS_RC(ctx);
+
+  // bail if this bind() is not one we're instrumenting
+  bind_syscall_args_t *args;
+  args = bpf_map_lookup_elem(&pending_bind, &tid);
+
+  log_debug("kretprobe/bind: for tid %d and return value: %d\n");
+
+  if (args == NULL) {
+    log_debug("kretprobe/bind: was not a UDP bind, will not process\n");
+    return 0;
+  }
+
+  if (ret != 0) {
+    log_debug("kretprobe/bind: return %d for tid %d\n", tid, ret);
+    return 0;
+  }
+
+
+  __u16 sin_port = args->port;
+  __u8 port_state = PORT_LISTENING;
+  bpf_map_update_elem(&udp_port_bindings, &sin_port, &port_state, BPF_ANY);
+  log_debug("kretprobe/bind: bound UDP port  %d\n", sin_port);
+
+  return 0;
+}
+
+// used for capturing UDP sockets that are bound
+SEC("kprobe/sys_socket")
+int kprobe__sys_socket(struct pt_regs* ctx) {
+    int domain = PT_REGS_PARM1(ctx);
+    int type = PT_REGS_PARM2(ctx);
+    __u64 tid = bpf_get_current_pid_tgid();
+
+    // figuring out if the socket being constructed is UDP. We will call
+    // a socket UDP if it is in the AF_INET or AF_INET6 domain. And
+    // the type is SOCK_DGRAM.
+    __u8 pending_udp = 0;
+    if ((domain & (AF_INET | AF_INET6)) > 0 && (type & SOCK_DGRAM) > 0) {
+        pending_udp = 1;
+    }
+
+    if (pending_udp == 0) {
+        log_debug("kprobe sys_socket: got a socket() call, but was not for UDP with tid=%d", tid);
+        return 0;
+    }
+
+    log_debug("kprobe sys_socket: started a UDP socket for tid=%d", tid);
+    __u8 x = 1;
+    bpf_map_update_elem(&pending_sockets, &tid, &x, BPF_ANY);
+
+    return 0;
+}
+
+// used in combination with the kprobe for sys_socket to find file descriptors for UDP sockets that have not
+// yet been "binded".
+SEC("kretprobe/sys_socket")
+int kretprobe__sys_socket(struct pt_regs* ctx) {
```

Comment:
Do we need to do any error handling here, say if `sys_socket` returned -1?  Would we want to clean up `pending_sockets` for that case?

Reply:
Nice catch, yes

Reply:
Done, I think this turned out to not be  technically necessary as it would require `bind(-1, addr)` to succeed, but it's good to keep the maps clean.